crates/rpc/rpc/src/reth.rs

```diff
@@ -0,0 +1,108 @@
+    /// The provider that can interact with the chain.
+    pub fn provider(&self) -> &Provider {
+        &self.inner.provider
+    }
+
+    /// Create a new instance of the [RethApi]
+    pub fn new(provider: Provider) -> Self {
+        let inner = Arc::new(RethApiInner { provider });
+        Self { inner }
+    }
+}
+
+impl<Provider> RethApi<Provider>
+where
+    Provider: AccountChangeReader + BlockReaderIdExt + StateProviderFactory + 'static,
+{
+    #[allow(unused)]
+    fn account_changes_in_block(
+        &self,
+        block_id: impl Into<BlockId>,
+    ) -> EthResult<HashMap<Address, Option<Account>>> {
+        let block_id = block_id.into();
+        let Some(block_number) = self.provider().block_number_for_id(block_id)? else {
+            return Err(EthApiError::UnknownBlockNumber)
+        };
+
+        let state = self.provider().state_by_block_id(block_id)?;
+        let accounts_before = self.provider().account_changes_in_block(block_number)?;
+        let hash_map = accounts_before.iter().try_fold(
+            HashMap::new(),
+            |mut hash_map, account_before| -> Result<_> {
+                let account = state.basic_account(account_before.address)?;
+                hash_map.insert(account_before.address, account);
+                Ok(hash_map)
+            },
+        )?;
+        Ok(hash_map)
+    }
+
+    fn balance_changes_in_block(
+        &self,
+        block_id: impl Into<BlockId>,
+    ) -> EthResult<HashMap<Address, Option<U256>>> {
+        let block_id = block_id.into();
+        let Some(block_number) = self.provider().block_number_for_id(block_id)? else {
+            return Err(EthApiError::UnknownBlockNumber)
+        };
+
+        let state = self.provider().state_by_block_id(block_id)?;
+        let accounts_before = self.provider().account_changes_in_block(block_number)?;
+        let hash_map = accounts_before.iter().try_fold(
+            HashMap::new(),
+            |mut hash_map, account_before| -> Result<_> {
+                let current_balance = state.account_balance(account_before.address)?;
+                let prev_balance = account_before.info.map(|info| info.balance);
+                if current_balance != prev_balance {
+                    hash_map.insert(account_before.address, current_balance);
+                }
+                Ok(hash_map)
+            },
+        )?;
+        Ok(hash_map)
+    }
+}
+
+impl<Provider> std::fmt::Debug for RethApi<Provider> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RethApi").finish_non_exhaustive()
+    }
+}
+
+#[async_trait]
+impl<Provider> RethApiServer for RethApi<Provider>
+where
+    Provider: AccountChangeReader + BlockReaderIdExt + StateProviderFactory + 'static,
+{
+    /// Handler for `reth_getBalancesChangesInBlock`
+    fn reth_get_balance_changes_in_block(
+        &self,
+        block_id: BlockId,
+    ) -> RpcResult<Option<HashMap<Address, Option<U256>>>> {
```

Comment:
```suggestion
    ) -> RpcResult<Option<HashMap<Address, U256>>> {
```

I think I probably will go with `0` for the destructed accounts as you suggested, as otherwise the entry will be `null` in the JSON probably differs from what erigon does, but I'll check over there too.

Reply:
Ok removed `Option` in https://github.com/paradigmxyz/reth/pull/3768/commits/a29d297886037be93bfe2a8199c85cde50a4285d.  I also removed the `Option<HashMap<...` in favor of always returning an empty HashMap, as I think this is the difference between returning `{}` and `null`, and `{}` seems more appropriate (the only time I think this even happens is maybe on a genesis block (and only then if the chain has no initial balances), or maybe on a chain w/ non-standard consensus rules that allows a new empty block w/o any coinbase rewards).